### PR TITLE
Add TARRAY_CREATE_WITH_CAPACITY

### PR DESCRIPTION
**Pull request by dcristoloveanu**

### CODE REVIEW

File: devdoc/tarray_requirements.md [hunk 1]
```diff
@@ -54,7 +54,7 @@ typedef struct TARRAY_STRUCT_T_TAG
 #define TARRAY_TYPE_DECLARE(T)
 ```
 
-`TARRAY_TYPE_DECLARE(T)` is a macro to be used in a header declaration. It introduces 2 APIs (as MOCKABLE_FUNCTIONS):
+`TARRAY_TYPE_DECLARE(T)` is a macro to be used in a header declaration. It introduces 3 APIs (as MOCKABLE_FUNCTIONS):
```
Comment (anporumb) [inline devdoc/tarray_requirements.md:57]: >3 [](http://example.com/codeflow?start=86&length=1)

accurante. I like that.

File: inc/c_util/tarray_ll.h [hunk 2]
```diff
@@ -99,6 +105,51 @@ TARRAY_LL(T) TARRAY_LL_CREATE(C)(void)
...
+        }                                                                                                                   \
+        else                                                                                                                \
+        {                                                                                                                   \
+            /* Codes_SRS_TARRAY_01_003: [ TARRAY_CREATE_WITH_CAPACITY(T) shall call malloc_2 to allocate capacity entries for result->arr. ]*/ \
+            result->arr = malloc_2(capacity,sizeof(T));                                                                     \
+            if(result->arr == NULL)                                                                                         \
+            {                                                                                                               \
+                /* Codes_SRS_TARRAY_01_005: [ If there are any failures then TARRAY_CREATE_WITH_CAPACITY(T) shall fail and return NULL. ]*/ \
+                LogError("failure in malloc(1 * sizeof(" MU_TOSTRING(T) ")=%zu)",                                           \
```
Comment (anporumb) [inline inc/c_util/tarray_ll.h:136]: > malloc(1 * sizeof(" MU_TOSTRING(T) ")=%zu)",   [](http://example.com/codeflow?start=36&length=47)

that is not what fails.

File: inc/c_util/tarray_ll.h [hunk 3]
```diff
@@ -99,6 +105,51 @@ TARRAY_LL(T) TARRAY_LL_CREATE(C)(void)
     return result;                                                                                                          \
 }
 
+/*introduces a function definition for tarray_create_with_capacity*/
+#define TARRAY_LL_CREATE_WITH_CAPACITY_DEFINE(C, T)                                                                         \
+TARRAY_LL(T) TARRAY_LL_CREATE_WITH_CAPACITY(C)(uint32_t capacity)                                                           \
```
Comment (anporumb) [inline inc/c_util/tarray_ll.h:110]: >TARRAY_LL_CREATE_WITH_CAPACITY [](http://example.com/codeflow?start=13&length=30)

can TARRAY_LL_CREATE be then replaced by a default call to TARRAY_LL_CREATE_WITH_CAPACITY where capacity is "1"?

Less code...

File: inc/c_util/tarray_ll.h [hunk 4]
```diff
@@ -99,6 +105,51 @@ TARRAY_LL(T) TARRAY_LL_CREATE(C)(void)
...
+            else                                                                                                            \
+            {                                                                                                               \
+                result->capacity = capacity;                                                                                \
+                /* Codes_SRS_TARRAY_01_004: [ TARRAY_CREATE_WITH_CAPACITY(T) shall succeed and return a non-NULL value. ]*/ \
+                goto allok;                                                                                                 \
+            }                                                                                                               \
+            THANDLE_FREE(TARRAY_TYPEDEF_NAME(C))(result);                                                                   \
+        }                                                                                                                   \
+        result = NULL;                                                                                                      \
```
Comment (anporumb) [inline inc/c_util/tarray_ll.h:147]: >result = NULL; [](http://example.com/codeflow?start=8&length=14)

if line 125 executes, then this line executes and sets "result" to NULL (again).

Move this line between 145, 146.

File: inc/c_util/tarray_ll.h [hunk 5]
```diff
@@ -99,6 +105,51 @@ TARRAY_LL(T) TARRAY_LL_CREATE(C)(void)
...
 
+/*introduces a function definition for tarray_create_with_capacity*/
+#define TARRAY_LL_CREATE_WITH_CAPACITY_DEFINE(C, T)                                                                         \
+TARRAY_LL(T) TARRAY_LL_CREATE_WITH_CAPACITY(C)(uint32_t capacity)                                                           \
+{                                                                                                                           \
+    TARRAY_TYPEDEF_NAME(T)* result;                                                                                         \
+    if (capacity == 0)                                                                                                      \
+    {                                                                                                                       \
+        /* Codes_SRS_TARRAY_01_001: [ If capacity is 0, TARRAY_CREATE_WITH_CAPACITY(T) shall fail and return NULL. ]*/      \
+        LogError("Invalid argumenets: uint32_t capacity=%" PRIu32 "", capacity);                                            \
```
Comment (anporumb) [inline inc/c_util/tarray_ll.h:116]: >argumenets [](http://example.com/codeflow?start=26&length=10)

I suppose these are some sort of juicy HARIBOes.

File: tests/tarray_ut/tarray_ut.c [hunk 6]
```diff
@@ -196,6 +201,131 @@ TEST_FUNCTION(TARRAY_of_THANDLE_can_be_created)
...
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG)); /*struct*/
+    STRICT_EXPECTED_CALL(malloc_2(1, sizeof(uint32_t))); /*inner array*/
+
+    ///act
+    TARRAY(uint32_t) arr_of_uint32_t = TARRAY_CREATE_WITH_CAPACITY(uint32_t)(1);
+
+    ///assert
+    ASSERT_IS_NOT_NULL(arr_of_uint32_t);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    arr_of_uint32_t->arr[0] = 4; /*assignment is possible*/
```
Comment (anporumb) [inline tests/tarray_ut/tarray_ut.c:234]: >/*assignment is possible*/ [](http://example.com/codeflow?start=33&length=26)

noice

File: tests/tarray_ut/tarray_ut.c [hunk 7]
```diff
@@ -196,6 +201,131 @@ TEST_FUNCTION(TARRAY_of_THANDLE_can_be_created)
...
+        THANDLE_INITIALIZE(A_TEST)(&arr_of_a->arr[i], a);
+        THANDLE_ASSIGN(A_TEST)(&arr_of_a->arr[i], NULL);
+        THANDLE_ASSIGN(A_TEST)(&a, NULL);
+    }
+
+    // cleanup
+    TARRAY_ASSIGN(THANDLE(A_TEST))(&arr_of_a, NULL);
+}
+
+/* TARRAY_ENSURE_CAPACITY(T) */
```
Comment (anporumb) [inline tests/tarray_ut/tarray_ut.c:327]: >TARRAY_ENSURE_CAPACITY [](http://example.com/codeflow?start=3&length=22)

consider adding at least 1 test for an array that is created with initial capacity and AFTER it is resized.

"should not fail".

**Review by anporumb — APPROVED** [review]
:shipit:

File: inc/c_util/tarray_ll.h (hunk 2 above)
Comment (dcristoloveanu) [reply inc/c_util/tarray_ll.h:136]: Fixed.

File: inc/c_util/tarray_ll.h (hunk 5 above)
Comment (dcristoloveanu) [reply inc/c_util/tarray_ll.h:116]: No more HARIBOs!

File: inc/c_util/tarray_ll.h (hunk 4 above)
Comment (dcristoloveanu) [reply inc/c_util/tarray_ll.h:147]: So good to have tests for these small refactors.

File: inc/c_util/tarray_ll.h (hunk 3 above)
Comment (dcristoloveanu) [reply inc/c_util/tarray_ll.h:86]: Sure.

**Pull request merged**